core/domain/exp_services_test.py

```diff
@@ -1987,7 +1987,7 @@ def test_export_to_zip_file(self):
 
         self.assertEqual(zf.namelist(), ['A title.yaml'])
         self.assertEqual(
-            zf.open('A title.yaml').read(), self.SAMPLE_YAML_CONTENT)
+            zf.open('A title.yaml').read(), self.SAMPLE_YAML_CONTENT) # pylint: disable=disallowed-function-calls
```

Comment:
This should not be disallowed, right? We are only disallowing open(...) function.

Reply:
Do you mean we should only be disregarding the python builtin function `open()`, rather than the attribute `zf.open()`? The checker currently disallows attributes with disallowed names, which includes `zf.open()` (rather than just disallowing `open()`).

If so, then there might be some changes required for determining whether we want to disallow attributes or just base functions in how we make the config. One way I can think of is adding some sort of special character in the beginning of the config entry (like `*attr=>new_attr`). If I misinterpreted before and we only want to disallow base functions, then I can change the logic.

Reply:
> Do you mean we should only be disregarding the python builtin function open(), rather than the attribute zf.open()? 

Yes! In general, this is a refactor/restructure PR and we don't expect this PR to changes any old functionality. [Note: Earlier we were only blocking the global open calls.]

Is there any issue with having open=>python_utils.open will only raise the issue if someone uses global open function and something.open=>python_utils.open will only raise the issue if someone uses something.open function? will it be possible to find the module name from the node in the visit_call function? [I'm not sure why we need a special character to represent this?]

Reply:
1. That makes sense. However, there's some functionality that we can't exactly capture by simply using a list of exact string replacements, such as the regex [`r'\.put_async\('` that was used in `general_purpose_linter`](https://github.com/oppia/oppia/pull/11557/files#diff-2ec95fb28a886d8e7cf26b726170551779907cbe7dde6bdd0f268076d70f4682L588-L600). What would we do in this case?

2. [**Please correct me if I'm wrong**] I think earlier in this PR you mentioned that I should not use the astroid nodes' [`as_string()` method](http://pylint.pycqa.org/projects/astroid/en/latest/api/base_nodes.html#astroid.node_classes.NodeNG.as_string). I think the complexity of this checker would rise drastically if we need to differentiate between attributes and global calls while strictly using astroid nodes. For example, the config `a.b.c=>d,a.b.c=>e` would require some trie like structure to store (`a->b->c->[d*,e*]` where `*`s represent leaves for lack of better representation). Then, checking in `visit_call` would require searching through the storage while doing type assertions on the underlying nodes (e.g. `b[Name].c[Attribute]()` is different than `a[Name].b[Attribute].c[Attribute]()`).

Apologies if I'm overcomplicating this and for the long PR. I think if we want to have the same functionality as we did in `general_purpose_linter`, it would be best to find a way to have the config use regex and to use the `as_string()` method on the astroid Call node to do the checking. Can you think of any other ways / are there simplifications we can make?

Reply:
> I should not use the astroid nodes' as_string() method.

Yeah, that's true we try to avoid string/regex checks wherever possible. I agree with you that this needs to be done via as string (and we may have to accept function-call-regex from .pylintrc and do some regex check for the as_string text for the cases like put_async.)

I'm fine with using as_string for this checker and if required you can change it to regex function-call pattern in the future (if you are finding it difficult to do in this PR!).